### PR TITLE
Updated the AddPolicyRegistry extension methods to use TryAdd and TryAddEnumerable internally

### LINKED ISSUE #47842: Add TryAddPolicyRegistry extension methods for IPolicyRegistry
## Background and Motivation
The existing ```AddPolicyRegistry()``` extension methods are registering the ```IPolicyRegistry``` as singleton to the ```IServiceCollection```. Any subsequent invocation of ```AddPolicyRegistry()``` will replace the already registered instance.

The goal of the proposed extension methods: ```TryAddPolicyRegistry()``` is to make sure the ```IPolicyRegistry``` is registered only once.
Any subsequent attempt to add it to DI will not replace the already registered instance. With this idempotent API it will be a bit easier to add policies to the same policy registry. The client code will still have to get the current instance from the ```IServiceCollection``` or create a new instance, but after that will be able to safely attempt to add it back to ```IServiceCollection```.

## Proposed API

The proposal is to add "TryAdd" alternatives to the existing ```AddPolicyRegistry``` APIs.
The implementation logic should not be changed, except for the usage of ```TryAddSingleton``` instead of ```AddSingleton```.

```diff
namespace Microsoft.Extensions.DependencyInjection;

public static class PollyServiceCollectionExtensions
{
+    public static IPolicyRegistry<string> TryAddPolicyRegistry(this IServiceCollection services);
+    public static IPolicyRegistry<string> TryAddPolicyRegistry(this IServiceCollection services, IPolicyRegistry<string> registry);
+    public static IServiceCollection TryAddPolicyRegistry(this IServiceCollection services, Action<IServiceProvider, IPolicyRegistry<string>> configureRegistry);
}
```

## Usage Examples

We have a client code method that registers HTTP clients in the ```IServiceCollection```:
```C#
public static IHttpClientBuilder AddHttpClientWithFaultHandling<TInterface, TImplementation>(this IServiceCollection services);
```
This method adds a set of policies (fallback, wait and retry, circuit breaker, timeout, logging, etc.) to the newly registered HTTP client.
This method will potentially be invoked multiple times in an application (once per HTTP client type).

The circuit breaker policy is tricky. It needs to be put in a ```PolicyRegistry```. Therefore the client code gets an instance of the ```PolicyRegistry``` from the ```IServiceCollection``` if it was already registered, or it creates a new instance:
```C#
private static IConcurrentPolicyRegistry<string> GetExistingOrNewPolicyRegistry(IServiceCollection services) =>
        services.BuildServiceProvider().GetService<IConcurrentPolicyRegistry<string>>() ?? new PolicyRegistry();
```

After getting the ```PolicyRegistry``` instance, this client code adds a new instance of the circuit breaker policy into the Policy Registry. It is assumed that each circuit breaker policy is different per HTTP client registration, therefore it can't be reused.
```C#
policyRegistry.AddOrUpdate($"{typeof(TImplementation).FullName}__CircuitBreakerPolicy", circuitBreakerPolicy, (x, y) => circuitBreakerPolicy);
```

With the help of the proposed APIs, the client code safely tries to add the ```PolicyRegistry``` to the ```IServiceCollection```. It should only be added on the 1st invocation. All subsequent invocations should recognize that the ```PolicyRegistry``` singleton is already registered and skip the action.

After the HTTP client is registered, the client code chains a set of policies to it. The circuit breaker policy is added using the ```AddPolicyHandlerFromRegistry``` method:
```C#
.AddPolicyHandlerFromRegistry($"{typeof(TImplementation).FullName}__CircuitBreakerPolicy")
```

### PR DESCRIPTION
**Pull request by Arclight3**
# Updated the AddPolicyRegistry extension methods to use TryAdd and TryAddEnumerable internally

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes:
Register a single instance for each ```IPolicyRegistry``` implementation type (concrete class).

## Description

When the current```AddPolicyRegistry()``` methods are invoked multiple times they replace the already registered instances.

The goal of these changes is to register ```IPolicyRegistry``` only once for each implementation type. Any subsequent attempts should not replace the already registered instances.

More details (e.g. the scenario where this change is useful) are found in this issue: https://github.com/dotnet/aspnetcore/issues/47842.

### CODE REVIEW

**Comment by captainsafia** [comment]
@Arclight3 Thanks for opening this PR! Since it introduces a new API, we'll have to take it through our API review process ([ref](https://github.com/dotnet/aspnetcore/blob/main/docs/APIReviewProcess.md)). You can file the API review issue [here](https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&template=30_api_proposal.md&title=).

**Comment by Arclight3** [comment]
Thank you @captainsafia. I added a new API proposal as you suggested: https://github.com/dotnet/aspnetcore/issues/47842.

File: src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -105,4 +106,101 @@ public static IServiceCollection AddPolicyRegistry(this IServiceCollection servi
...
+        {
+            throw new ArgumentNullException(nameof(services));
+        }
+
+        if (registry == null)
+        {
+            throw new ArgumentNullException(nameof(registry));
+        }
+
+        services.TryAddSingleton<IPolicyRegistry<string>>(registry);
```
Comment (amcasey) [inline src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs:157]: I have very little area knowledge, but should these additions be all-or-nothing?  Does it make sense to add the `IReadOnlyPolicyRegistry` if the `IPolicyRegistry` wasn't found (e.g. because there was already one registered)?

Comment (Arclight3) [reply src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs:157]: I also don't have much knowledge of this area, but reading the existing code I saw that the policy registry must always be registered for all 3 interfaces (```IConcurrentPolicyRegistry```, ```IPolicyRegistry``` and ```IReadOnlyPolicyRegistry```).
Therefore, I would think there should never be a case where only one interface is registered. As you said, it's all or nothing.

In any case, I think the ```Add``` and ```TryAdd``` methods should register the same objects in the end. If we decide that the ```TryAdd```s should not be "all-or-nothing", we should also update the ```Add``` methods.

**Comment by Arclight3** [comment]
Hi,

This is my first time trying to commit something to the aspnetcore project.
Can anyone please tell me how is a new API added properly so the build passes?
If I add the API in PublicAPI.Shipped.txt the build fails. If I don't it also fails :)

**Comment by captainsafia** [comment]
> Hi,
> 
> This is my first time trying to commit something to the aspnetcore project. Can anyone please tell me how is a new API added properly so the build passes? If I add the API in PublicAPI.Shipped.txt the build fails. If I don't it also fails :)

You should typically apply the updates to the PublicApi.*Unshipped*.txt files. If you are editing in Visual Studio, you should be able to right-click on the diagnostic in VS and see an option to add the new API to text file.

**Comment by amcasey** [comment]
@Arclight3 API review [suggested](https://github.com/dotnet/aspnetcore/issues/47842#issuecomment-1544790243) a different approach.  Do you want to update this PR or start fresh?

**Comment by Arclight3** [comment]
> @Arclight3 API review [suggested](https://github.com/dotnet/aspnetcore/issues/47842#issuecomment-1544790243) a different approach. Do you want to update this PR or start fresh?

Thanks for the update! I will update the current PR a bit later today 😄

**Comment by BrennanConroy** [comment]
You have test issues caused by the PR, rerunning the CI wont fix that.

**Comment by Arclight3** [comment]
> You have test issues caused by the PR, rerunning the CI wont fix that.

Yep :)

I saw initially "Helix x64" and I thought it's another platform like Linux. Since the build passed on Linux I thought this is something specific to this platform (something transient) and another build might solve it.

Also, this is my first PR and I am not familiar with these specific pipelines. At a first look at the error messages it's not so obvious that some tests failed (and which tests).
I think I found them now and I will try to fix them :)

**Comment by Arclight3** [comment]
Fixed 😄

**Comment by Arclight3** [comment]
Please review again the PR and let me know if I need to change something else.

**Comment by captainsafia** [comment]
@Arclight3 My main feedback was around adding test coverage for this in https://github.com/dotnet/aspnetcore/blob/main/src/HttpClientFactory/Polly/test/DependencyInjection/PollyHttpClientBuilderExtensionsTest.cs.

**Comment by Arclight3** [comment]
Alright. I will add tomorrow more tests to cover the new use cases.
Thanks for the feedback! 😄

File: src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -30,17 +32,18 @@ public static IPolicyRegistry<string> AddPolicyRegistry(this IServiceCollection
         // single instance registered using all the interfaces.
         var registry = new PolicyRegistry();
 
-        services.AddSingleton<IConcurrentPolicyRegistry<string>>(registry);
-        services.AddSingleton<IPolicyRegistry<string>>(registry);
-        services.AddSingleton<IReadOnlyPolicyRegistry<string>>(registry);
+        services.TryAddEnumerable(ServiceDescriptor.Singleton<IPolicyRegistry<string>>(registry));
```
Comment (amcasey) [inline src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs:35]: Out of curiosity, why change the order?

Comment (Arclight3) [reply src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs:35]: > Out of curiosity, why change the order?

There are 3 extension methods that always register the same 3 types and I thought it would be a bit better (more consistent and intention clear) to have the same registration order in all methods: ```IPolicyRegistry``` then ```IReadOnlyPolicyRegistry``` then ```IConcurrentPolicyRegistry```. I saw this order already being used in the 2nd extension method and it made the most sense to me.
I tried not to be careless and I thought if changing the order would break something but I couldn't find an issue.

Comment (amcasey) [reply src/HttpClientFactory/Polly/src/DependencyInjection/PollyServiceCollectionExtensions.cs:35]: Makes sense to me.  Thanks for the explanation!

**Comment by Arclight3** [comment]
While writing the unit tests I faced a little problem.
The ```AddPolicyRegistry``` methods return the registry object to the caller.
Since we want to have a single policy registry in the service collection it makes sense to return the existing instance in case it is found, otherwise return a new policy registry instance or the one provided by the user.

In the first extension method we have no problem to achieve it:
```C#
public static IPolicyRegistry<string> AddPolicyRegistry(this IServiceCollection services)
{
    if (services == null)
    {
        throw new ArgumentNullException(nameof(services));
    }

    // If not already registered create an empty registry, register and return it as an instance.
    // This is the best way to get a single instance registered using all the interfaces.
    var registry = services.BuildServiceProvider().GetService<IPolicyRegistry<string>>();
    if (registry == null)
    {
        registry = new PolicyRegistry();
    }

    services.TryAddEnumerable(ServiceDescriptor.Singleton<IPolicyRegistry<string>>(registry));
    services.TryAddEnumerable(ServiceDescriptor.Singleton<IReadOnlyPolicyRegistry<string>>(registry));

    if (registry is IConcurrentPolicyRegistry<string> concurrentRegistry)
    {
        services.TryAddEnumerable(ServiceDescriptor.Singleton<IConcurrentPolicyRegistry<string>>(concurrentRegistry));
    }

    return registry;
}
```
This test scenario would pass:
```C#
 // Arrange
var serviceCollection = new ServiceCollection();
var existingRegistry = new PolicyRegistry();

serviceCollection.AddSingleton<IPolicyRegistry<string>>(existingRegistry);
serviceCollection.AddSingleton<IReadOnlyPolicyRegistry<string>>(existingRegistry);
serviceCollection.AddSingleton<IConcurrentPolicyRegistry<string>>(existingRegistry);

// Act
var registry = serviceCollection.AddPolicyRegistry();

// Assert
Assert.Same(existingRegistry, registry);
```

But on the 2nd method we have a conflict. We receive a registry instance as a parameter (possibly with some policies in it):
```C#
public static IPolicyRegistry<string> AddPolicyRegistry(this IServiceCollection services, IPolicyRegistry<string> registry)
{
    if (services == null)
    {
        throw new ArgumentNullException(nameof(services));
    }

    if (registry == null)
    {
        throw new ArgumentNullException(nameof(registry));
    }

    services.TryAddEnumerable(ServiceDescriptor.Singleton<IPolicyRegistry<string>>(registry));
    services.TryAddEnumerable(ServiceDescriptor.Singleton<IReadOnlyPolicyRegistry<string>>(registry));

    if (registry is IConcurrentPolicyRegistry<string> concurrentRegistry)
    {
        services.TryAddEnumerable(ServiceDescriptor.Singleton<IConcurrentPolicyRegistry<string>>(concurrentRegistry));
    }

    return registry;
}
```
In case we already have a policy registered what should we do?
1. Should we copy the policies from the input parameter registry into the existing registry and then return the existing policy
2. We don't register anything from the input parameter registry if we already have an instance registered (because we use the TryAdd methods) and we return the existing policy
3. We don't register anything from the input parameter registry if we already have an instance registered (because we use the TryAdd methods) and we return the input parameter registry

I tend to like option 1 because is more robust and doesn't allow any misuses, but we may end up building something too complicated..

For the 3rd method we don't have a conflict. It's similar to the 1st. We get the existing instance or create a new one before configuring it:
```C#
public static IServiceCollection AddPolicyRegistry(this IServiceCollection services, Action<IServiceProvider, IPolicyRegistry<string>> configureRegistry)
{
    if (services == null)
    {
        throw new ArgumentNullException(nameof(services));
    }

    if (configureRegistry == null)
    {
        throw new ArgumentNullException(nameof(configureRegistry));
    }

    // Create an empty registry, configure it and register it as an instance.
    // This is the best way to get a single instance registered using all the interfaces.
    services.TryAddSingleton(serviceProvider =>
    {
        var registry = services.BuildServiceProvider().GetService<IPolicyRegistry<string>>();
        if (registry == null)
        {
            registry = new PolicyRegistry();
        }

        configureRegistry(serviceProvider, registry);

        return registry;
    });

    services.TryAddEnumerable(ServiceDescriptor.Singleton<IPolicyRegistry<string>, PolicyRegistry>(serviceProvider => serviceProvider.GetRequiredService<PolicyRegistry>()));
    services.TryAddEnumerable(ServiceDescriptor.Singleton<IReadOnlyPolicyRegistry<string>, PolicyRegistry>(serviceProvider => serviceProvider.GetRequiredService<PolicyRegistry>()));
    services.TryAddEnumerable(ServiceDescriptor.Singleton<IConcurrentPolicyRegistry<string>, PolicyRegistry>(serviceProvider => serviceProvider.GetRequiredService<PolicyRegistry>()));

    return services;
}
```

Please share your thoughts 😄

**Comment by Arclight3** [comment]
Regarding my prior message, I updated the implementation to move (where applicable) the policies from the new registry to the existing registry instance and return the existing registry instance.
In case of policies with duplicate key name I thought it's best to mimic the behavior of manually registering them which is to throw an exception.

I also added unit tests to cover all the 'existing vs new' registry scenarios I could think of 😄 

Please review the latest changes and let me know what you think of this implementation or if you think we should implement it differently.

**Comment by amcasey** [comment]
@Arclight3 thanks for all your hard work on this!  I just wanted to let you know that responses might be a little slow since a lot of us are heading into a long weekend, but you haven't been forgotten. :smile:

**Comment by Arclight3** [comment]
Great! Thanks for the heads up @amcasey! 😄

**Comment by amcasey** [comment]
> Regarding my prior message, I updated the implementation to move (where applicable) the policies from the new registry to the existing registry instance and return the existing registry instance. 

Well, it absolutely must return whichever instance you would get upon retrieving it from the collection, so that part sounds correct.  I haven't decided about the merging yet.

> In case of policies with duplicate key name I thought it's best to mimic the behavior of manually registering them which is to throw an exception.

If we go with merging, this sounds right.

> In the last method (with the configuration delegate) there was actually a problem and I had to make a compromise. If the method was called multiple times only the last delegate was invoked, which meant all prior configurations (and their policies) were replaced. 

Tell me more about when the delegates do / need to run?  Do they have to run before the DI container is sealed?  It seems like the policy registry can be retrieved and mutated after that.

> In order to invoke all configurations I had to build the services collection with whatever services were registered at that time. This means that if a configuration delegate depends on some services that will be registered later, the delegate will throw. This is a breaking change but I thought it's more important not to lose registered policies.

I didn't fully understand this, but it sounds iffy.  I think you're saying that you want the delegates to run after the DI container is sealed, but they're presently running before and that causes issues?

**Comment by Arclight3** [comment]
> > Tell me more about when the delegates do / need to run? Do they have to run before the DI container is sealed? It seems like the policy registry can be retrieved and mutated after that.

In the current version of the ```AddPolicyRegistry``` method, the delegates run after the container is sealed. They run when the services are retrieved from the container.

**Pros**: Because delegates runs after the container is sealed, the delegates and their dependencies can be registered in any order.
**Cons**: Only the last registered delegate for a given type is invoked.

Example:
```C#
 var serviceCollection = new ServiceCollection();

// Act 1
serviceCollection.AddPolicyRegistry((serviceProvider, registry) =>
{
    registry.Add<IAsyncPolicy<HttpResponseMessage>>("retryPolicy1", RetryPolicy);
});

// Act 2
serviceCollection.AddPolicyRegistry((serviceProvider, registry) =>
{
    registry.Add<IAsyncPolicy<HttpResponseMessage>>("retryPolicy2", RetryPolicy);
});

// Seal the container
var services = serviceCollection.BuildServiceProvider();

// Request a service
var registry = services.GetService<IPolicyRegistry<string>>();
```
The delegate is invoked  at this line: ```var registry = services.GetService<IPolicyRegistry<string>>();```.
Only the last delegate (from Act 2) is invoked. The delegate from Act 1 was replaced the moment we registered the delegate from Act 2; It's gone along with the 'retryPolicy1' policy. The registry object contains only the 'retryPolicy2' policy.

This is the logic that invokes the delegate:
```C#
services.AddSingleton(serviceProvider =>
{
    var registry = new PolicyRegistry();

    configureRegistry(serviceProvider, registry);

    return registry;
});
```
It's registering an implementation factory that will be executed only once when the service is requested.
If this is executed multiple times the implementation factory is replaced.

I also tested the ```TryAddSingleton``` but it behaves similarly. It invokes only the 1st delegate (from Act 1) instead of the last delegate. It's still not invoking both.

> > I didn't fully understand this, but it sounds iffy. I think you're saying that you want the delegates to run after the DI container is sealed, but they're presently running before and that causes issues?

Ideally I would like to run all delegates after the container is sealed. But I didn't find a way to do so.

The solution I proposed is to run the delegates right when they are registered, before the container is sealed.
I replaced this part:
```C#
services.AddSingleton(serviceProvider =>
{
    var registry = new PolicyRegistry();

    configureRegistry(serviceProvider, registry);

    return registry;
});
```

with 

```C#
var serviceProvider = services.BuildServiceProvider();

// Get existing registry or an empty instance
var registry = serviceProvider.GetService<IPolicyRegistry<string>>();
if (registry == null)
{
    registry = new PolicyRegistry();
}

configureRegistry(serviceProvider, registry);
````

This way we don't lose any policy no matter how many delegates the client code will setup, but on the other hand the registration order of the delegate and it's dependencies now matter. If the dependencies are registered after the delegate the execution will fail.

Example:
```C#
var options = new PollyPolicyOptions
{
    PolicyName = "retrypolicy"
};

var serviceCollection = new ServiceCollection();

// Act 1
// This throws: System.InvalidOperationException:
// 'No service for type 'Microsoft.Extensions.DependencyInjection.PollyHttpClientBuilderExtensionsTest+PollyPolicyOptions' has been registered.'
serviceCollection.AddPolicyRegistry((serviceProvider, registry) =>
{
    // Required service
    var policyName = serviceProvider.GetRequiredService<PollyPolicyOptions>().PolicyName;
    registry.Add<IAsyncPolicy<HttpResponseMessage>>(policyName, RetryPolicy);
});

// Required service registered after delegate invocation
serviceCollection.AddSingleton(options);

var services = serviceCollection.BuildServiceProvider();

var registry = services.GetService<IPolicyRegistry<string>>();
```

**Comment by amcasey** [comment]
@Arclight3 Based on all the thought you've put into this and the points you raised (thanks again!), it seems like the only way to make this consistent across all the overloads and with existing APIs would be to actually modify some of the signatures.  Since we're not really interested in adding complexity to this area right now (or going through all the careful steps required to transitions consumers to new APIs), we think the best way forward is to update _only_ the parameterless overload.  

Our reasoning was that, generally speaking, someone calling `AddPolicyRegistry()` seems to be expressing their intention to ensure that _some_ policy registry is available without being too concerned about how it got there.  Changing it from an add-unconditionally operation to an add-if-missing operation should only confuse users who were intentionally trying to replace an existing policy registry with a blank one - which seems likely to be rare in practice.  So, updating just the one overload technically makes it inconsistent with the others, but not in a way that is likely to cause many problems.  Plus it solves your problem. :smile:

TL;DR: Let's make `AddPolicyRegistry()` try-add and the revert the changes to the other two overloads.

Does that seem reasonable?

**Comment by Arclight3** [comment]
@amcasey I agree with everything that you said and this solution works for me!
I'll update the PR soon.

Thank you for the support, guidance, and invaluable insights you provided throughout the process of my pull request!🤩

**Comment by amcasey** [comment]
Thanks!  This change LGTM, but I'd find it reassuring to hear from @captainsafia.

**Comment by Arclight3** [comment]
I synced the latest changes and successfully rerun the CI pipeline.

@captainsafia please have another look at the PR changes. Thanks! 😄

**Review by amcasey — APPROVED** [review]

**Comment by amcasey** [comment]
I talked to @captainsafia offline. :smile:

**Comment by amcasey** [comment]
Thanks again, @Arclight3!

**Comment by Arclight3** [comment]
Awesome! 🤩
Thank you as well!
